crates/gpui/src/platform/windows/window.rs

```diff
@@ -345,155 +329,210 @@ impl WindowsWindowInner {
-            if let Some(callback) = callbacks.input.as_mut() {
-                let ime_key = keystroke.ime_key.clone();
-                let event = KeyDownEvent {
-                    keystroke,
-                    is_held: true,
-                };
-
-                if callback(PlatformInput::KeyDown(event)) {
-                    CallbackResult::Handled { by_callback: true }
-                } else if let Some(mut input_handler) = self.input_handler.take() {
-                    if let Some(ime_key) = ime_key {
-                        input_handler.replace_text_in_range(None, &ime_key);
-                    }
-                    self.input_handler.set(Some(input_handler));
-                    CallbackResult::Handled { by_callback: true }
-                } else {
-                    CallbackResult::Handled { by_callback: false }
-                }
-            } else {
-                CallbackResult::Handled { by_callback: false }
-            }
+    fn parse_char_msg_keystroke(&self, wparam: WPARAM) -> Option<Keystroke> {
+        let src = [wparam.0 as u16];
+        let Ok(first_char) = char::decode_utf16(src).collect::<Vec<_>>()[0] else {
+            return None;
+        };
+        if first_char.is_control() {
+            None
         } else {
-            CallbackResult::Unhandled
+            Some(Keystroke {
+                modifiers: self.current_modifiers(),
+                key: first_char.to_lowercase().to_string(),
+                ime_key: Some(first_char.to_string()),
+            })
         }
     }
 
-    fn handle_keyup_msg(&self, wparam: WPARAM) -> CallbackResult {
-        let mut callbacks = self.callbacks.borrow_mut();
-        let keystroke = self.parse_key_msg_keystroke(wparam);
-        if let Some(keystroke) = keystroke {
-            if let Some(callback) = callbacks.input.as_mut() {
-                let event = KeyUpEvent { keystroke };
-                let by_callback = callback(PlatformInput::KeyUp(event));
-                CallbackResult::Handled { by_callback }
-            } else {
-                CallbackResult::Handled { by_callback: false }
-            }
-        } else {
-            CallbackResult::Unhandled
+    fn handle_syskeydown_msg(&self, message: u32, wparam: WPARAM, lparam: LPARAM) -> LRESULT {
+        // we need to call `DefWindowProcW`, or we will lose the system-wide `Alt+F4`, `Alt+{other keys}`
+        // shortcuts.
+        let Some(keystroke) = self.parse_syskeydown_msg_keystroke(wparam) else {
+            return unsafe { DefWindowProcW(self.hwnd, message, wparam, lparam) };
+        };
+        let Some(ref mut func) = self.callbacks.borrow_mut().input else {
+            return unsafe { DefWindowProcW(self.hwnd, message, wparam, lparam) };
+        };
+        let event = KeyDownEvent {
+            keystroke,
+            is_held: lparam.0 & (0x1 << 30) > 0,
+        };
+        if func(PlatformInput::KeyDown(event)) {
+            self.invalidate_client_area();
```

Comment:
I don't think all these invalidate client rects on input are necessary since we're supposed to be rendering at the display refresh rate :)

if we could remove them in all the keydown/keyup/char messages that would be great. unless they're needed for some other reason?
```suggestion
```

Reply:
Sometimes, even though the timer is set to trigger every 16ms, in some heavy scenarios, it might take 100ms for the timer to trigger, possibly due to scheduling and thread priority. So, scheduling a redraw here to prompt Zed to react as soon as possible.

Reply:
with `DCompositionWaitForCompositorClock` being used invalidating the rect won't trigger another message. I understand with the timer it's beneficial, but forcing an invalidate on input is more like a hack than a fix.